File: gsi/sysconfig/source/test/system-config-test.c
/*
 * Copyright 1999-2016 University of Chicago
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 * http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "globus_gsi_system_config.h"

int main(int argc, char *argv[])
{
    globus_result_t                     result = GLOBUS_SUCCESS;

    globus_module_activate(GLOBUS_GSI_SYSCONFIG_MODULE);

    if (argc < 2)
    {
        fprintf(stderr, "Error: %s cmd [args..]\n", argv[0]);
        exit(EXIT_FAILURE);
    }
    else if (strcmp(argv[1], "set_key_permissions") == 0
        && argc == 3)
    {
        result = GLOBUS_GSI_SYSCONFIG_SET_KEY_PERMISSIONS(argv[2]);
    }
    else if (strcmp(argv[1], "get_home_dir") == 0
        && argc == 2)
    {
        char                           *home_dir = NULL;

        result = GLOBUS_GSI_SYSCONFIG_GET_HOME_DIR(&home_dir);

        if (home_dir != NULL)
        {
            printf("%s\n", home_dir);
        }
        free(home_dir);
    }
    else if (strcmp(argv[1], "file_exists") == 0
        && argc == 3)
    {
        result = GLOBUS_GSI_SYSCONFIG_FILE_EXISTS(argv[2]);
    }
    else if (strcmp(argv[1], "dir_exists") == 0
        && argc == 3)
    {
        result = GLOBUS_GSI_SYSCONFIG_DIR_EXISTS(argv[2]);
    }
    else if (strcmp(argv[1], "check_keyfile") == 0
        && argc == 3)
    {
        result = GLOBUS_GSI_SYSCONFIG_CHECK_KEYFILE(argv[2]);
    }
    else if (strcmp(argv[1], "check_certfile") == 0
        && argc == 3)
    {
        result = GLOBUS_GSI_SYSCONFIG_CHECK_CERTFILE(argv[2]);
    }
    else if (strcmp(argv[1], "get_cert_dir") == 0
        && argc == 2)
    {
        char                           *cert_dir = NULL;

        result = GLOBUS_GSI_SYSCONFIG_GET_CERT_DIR(&cert_dir);

        if (cert_dir != NULL)
        {
            printf("%s\n", cert_dir);
        }
    }
    else if (strcmp(argv[1], "get_user_cert_filename") == 0
        && argc == 2)
    {
        char                           *user_cert_filename = NULL;
        char                           *user_key_filename = NULL;

        result = GLOBUS_GSI_SYSCONFIG_GET_USER_CERT_FILENAME(
                &user_cert_filename,
                &user_key_filename);

        if (user_cert_filename != NULL)
        {
            printf("%s\n", user_cert_filename);
        }
        if (user_key_filename != NULL)
        {
            printf("%s\n", user_key_filename);
        }

    }
    else if (strcmp(argv[1], "get_host_cert_filename") == 0
        && argc == 2)
    {
        char                           *host_cert_filename = NULL;
        char                           *host_key_filename = NULL;

        result = GLOBUS_GSI_SYSCONFIG_GET_HOST_CERT_FILENAME(
                &host_cert_filename,
                &host_key_filename);

        if (host_cert_filename != NULL)
        {
            printf("%s\n", host_cert_filename);
        }
        if (host_key_filename != NULL)
        {
            printf("%s\n", host_key_filename);
        }
    }
    else if (strcmp(argv[1], "get_service_cert_filename") == 0
        && argc == 3)
    {
        char                           *service_cert_filename = NULL;
        char                           *service_key_filename = NULL;

        result = GLOBUS_GSI_SYSCONFIG_GET_SERVICE_CERT_FILENAME(
                argv[2],
                &service_cert_filename,
                &service_key_filename);

        if (service_cert_filename != NULL)
        {
            printf("%s", service_cert_filename);
        }
        if (service_key_filename != NULL)
        {
            printf("%s", service_key_filename);
        }
    }
    else if (strcmp(argv[1], "get_proxy_filename") == 0
        && argc == 3)
    {
        int                             proxy_file_type = atoi(argv[3]);
        char                           *proxy_filename = NULL;

        result = GLOBUS_GSI_SYSCONFIG_GET_PROXY_FILENAME(
                &proxy_filename,
                (globus_gsi_proxy_file_type_t) proxy_file_type);

        if (proxy_filename != NULL)
        {
            printf("%s\n", proxy_filename);
        }
    }
    else if (strcmp(argv[1], "get_signing_policy_filename") == 0
        && argc == 4)
    {
        char                           *signing_policy_filename = NULL;
        char                           *ca_name_string = argv[1];

        globus_assert_string(0, "Test not implemented");
    }
    else if (strcmp(argv[1], "get_ca_cert_files") == 0
        && argc == 3)
    {
        globus_fifo_t                   ca_cert_list;

        result = GLOBUS_GSI_SYSCONFIG_GET_CA_CERT_FILES(
                argv[2],
                &ca_cert_list);

        if (result == GLOBUS_SUCCESS)
        {
            while (!globus_fifo_empty(&ca_cert_list))
            {
                char                   *cert = NULL;
                
                cert = globus_fifo_dequeue(&ca_cert_list);
                printf("%s\n", cert);
                free(cert);
            }
        }
    }
    else if (strcmp(argv[1], "get_current_working_dir") == 0
        && argc == 2)
    {
        char                           *working_dir = NULL;

        result = GLOBUS_GSI_SYSCONFIG_GET_CURRENT_WORKING_DIR(&working_dir);

        if (working_dir != NULL)
        {
            printf("%s\n", working_dir);
        }
    }
    else if (strcmp(argv[1], "make_absolute_path_for_filename") == 0
        && argc == 3)
    {
        char                           *absolute_path = NULL;

        result = GLOBUS_GSI_SYSCONFIG_MAKE_ABSOLUTE_PATH_FOR_FILENAME(
                argv[2],
                &absolute_path);

        if (absolute_path != NULL)
        {
            printf("%s\n", absolute_path);
        }
    }
    else if (strcmp(argv[1], "split_dir_and_filename") == 0
        && argc == 3)
    {
        char                           *dir_string = NULL;
        char                           *filename_string = NULL;

        result = GLOBUS_GSI_SYSCONFIG_SPLIT_DIR_AND_FILENAME(
                argv[2],
                &dir_string,
                &filename_string);

        if (dir_string != NULL)
        {
            printf("%s\n", dir_string);
        }
        if (filename_string != NULL)
        {
            printf("%s\n", filename_string);
        }
    }
    else if (strcmp(argv[1], "remove_all_owned_files") == 0
        && argc == 3)
    {
        result = GLOBUS_GSI_SYSCONFIG_REMOVE_ALL_OWNED_FILES(argv[2]);
    }
    else if (strcmp(argv[1], "is_superuser") == 0
        && argc == 2)
    {
        int                             is_superuser = -1;

        result = GLOBUS_GSI_SYSCONFIG_IS_SUPERUSER(&is_superuser);

        printf("%d\n", is_superuser);
    }
    else if (strcmp(argv[1], "get_user_id_string") == 0
        && argc == 2)
    {
        char                           *user_id_string = NULL;

        result = GLOBUS_GSI_SYSCONFIG_GET_USER_ID_STRING(&user_id_string);

        if (user_id_string != NULL)
        {
            printf("%s\n", user_id_string);
        }
    }
    else if (strcmp(argv[1], "get_username") == 0
        && argc == 2)
    {
        char                           *username = NULL;

        result = GLOBUS_GSI_SYSCONFIG_GET_USERNAME(
                &username);

        if (username != NULL)
        {
            printf("%s\n", username);
        }
    }
    else if (strcmp(argv[1], "get_proc_id_string") == 0
        && argc == 2)
    {
        char                           *proc_id_string = NULL;

        result = GLOBUS_GSI_SYSCONFIG_GET_PROC_ID_STRING(
                &proc_id_string);

        if (proc_id_string != NULL)
        {
            printf("%s\n", proc_id_string);
        }
    }
    else if (strcmp(argv[1], "get_gridmap_filename") == 0
        && argc == 2)
    {
        char                           *gridmap_filename = NULL;

        result = GLOBUS_GSI_SYSCONFIG_GET_GRIDMAP_FILENAME(
                &gridmap_filename);

        if (gridmap_filename != NULL)
        {
            printf("%s\n", gridmap_filename);
        }
    }
    else if (strcmp(argv[1], "get_authz_conf_filename") == 0
        && argc == 2)
    {
        char                           *authz_conf_filename = NULL;

        result = GLOBUS_GSI_SYSCONFIG_GET_AUTHZ_CONF_FILENAME(
                &authz_conf_filename);

        if (authz_conf_filename != NULL)
        {
            printf("%s\n", authz_conf_filename);
        }
    }
    else if (strcmp(argv[1], "get_unique_proxy_filename") == 0)
    {
        char                           *unique_filename = NULL;

        result = GLOBUS_GSI_SYSCONFIG_GET_UNIQUE_PROXY_FILENAME(
                &unique_filename);

        if (unique_filename != NULL)
        {
            printf("%s\n", unique_filename);
        }
    }
    else
    {

        fprintf(stderr, "Invalid arguments:");
        for (int i = 1; i < argc; i++)
        {
            fprintf(stderr, " %s", argv[i]);
        }
        fprintf(stderr, "\n");
    }

    if (result != GLOBUS_SUCCESS)
    {
        char                           *msg = NULL;

        msg = globus_error_print_friendly(globus_error_peek(result));

        fprintf(stderr, "ERROR: %s\n", msg ? msg : "Unknown");
        exit(EXIT_FAILURE);
    }
    exit(EXIT_SUCCESS);
}

File: gsi/sysconfig/source/test/system-config-test.pl
#! /usr/bin/perl
# Copyright 1999-2016 University of Chicago
# 
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
# 
# http://www.apache.org/licenses/LICENSE-2.0
# 
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

use strict;
use warnings;

use Fcntl ':mode';
use File::Path;
use File::Temp ':POSIX';
use Test::More;

my $system_config_test = "./system-config-test";

sub set_key_permissions {
    my $umask = umask(0);
    my $filename = tmpnam();
    my ($output, $rc, $mode);

    $output = qx($system_config_test set_key_permissions $filename);
    $rc = $? << 8;

    diag($output) if $output ne "";

    $mode = (stat($filename))[2] & (S_IRWXU|S_IRWXG|S_IRWXO);
    diag(sprintf "mode of new file is %o", $mode);

    ok($rc == 0 && ($mode & (S_IRWXG|S_IRWXO)) == 0, "set_key_permissions");
    unlink($filename);
    umask($umask);
}

sub get_home_dir {
    my $homedir = (getpwuid($<))[7];
    my ($rc, $output);

    $output = qx($system_config_test get_home_dir);
    $rc = $? << 8;

    $output =~ s/\n$//;

    diag($output) if $output ne "";

    ok($homedir eq $output, "get_home_dir");
}

sub file_exists_true {
    my ($fh, $filename) = File::Temp::tempfile();
    my ($rc, $output);

    print $fh "Hello\n";
    $fh->flush();

    $output = qx($system_config_test file_exists $filename);
    $rc = $? << 8;

    ok($rc == 0, "file_exists_true");
}

sub file_exists_false {
    my ($fh, $filename) = File::Temp::tempfile();
    my ($rc, $output);

    close($fh);
    unlink($filename);

    $output = qx($system_config_test file_exists $filename 2>&1);
    $rc = $? << 8;

    chomp($output);
    diag("Expect error message: $output");

    ok($rc != 0, "file_exists_false");
}

sub dir_exists_true {
    my ($dir) = File::Temp::tempdir(CLEANUP => 1);
    my ($rc, $output);


    $output = qx($system_config_test dir_exists $dir);
    $rc = $? << 8;

    ok($rc == 0, "dir_exists_true");
}

sub dir_exists_false {
    my ($dir) = File::Temp::tempdir();
    my ($rc, $output);

    rmdir $dir;

    $output = qx($system_config_test dir_exists $dir 2>&1);
    $rc = $? << 8;

    chomp($output);
    diag("Expect error message: $output");

    ok($rc != 0, "dir_exists_false");
}

sub check_keyfile_true {
    my $umask = umask(0);
    my $filename = tmpnam();
    my ($output, $rc, $mode, $fh);

    $output = qx($system_config_test set_key_permissions $filename);
    $rc = $? << 8;

    diag($output) if $output ne "";
    $mode = (stat($filename))[2] & (S_IRWXU|S_IRWXG|S_IRWXO);
    diag(sprintf "mode of new file is %o", $mode);

    open($fh, ">$filename");
    print $fh "Hello\n";
    close($fh);

    $output = qx($system_config_test check_keyfile $filename);

    $rc = $? << 8;

    ok($rc == 0, "check_keyfile_true");
    unlink($filename);
    umask($umask);
}

sub check_keyfile_false {
    my $umask = umask(0);
    my ($fh, $filename) = File::Temp::tempfile();
    my ($output, $rc, $mode);

    chmod 0644, $filename;
    print $fh "Hello\n";
    $fh->flush();

    $output = qx($system_config_test check_keyfile $filename 2>&1);
    $rc = $? << 8;

    chomp($output);
    diag("Expect error: $output");

    $mode = (stat($filename))[2] & (S_IRWXU|S_IRWXG|S_IRWXO);

    ok($rc != 0 && ($mode & (S_IRWXG|S_IRWXO)) != 0, "check_keyfile_false");
    unlink($filename);
    umask($umask);
}

sub check_certfile_true {
    my $umask = umask(0);
    my $filename = tmpnam();
    my ($output, $rc, $mode, $fh);

    $output = qx($system_config_test set_key_permissions $filename);
    $rc = $? << 8;

    diag($output) if $output ne "";
    $mode = (stat($filename))[2] & (S_IRWXU|S_IRWXG|S_IRWXO);
    diag(sprintf "mode of new file is %o", $mode);

    open($fh, ">$filename");
    print $fh "Hello\n";
    close($fh);

    $output = qx($system_config_test check_certfile $filename);

    $rc = $? << 8;

    ok($rc == 0, "check_certfile_true");
    unlink($filename);
    umask($umask);
}

sub check_certfile_false {
    my $umask = umask(0);
    my ($fh, $filename) = File::Temp::tempfile();
    my ($output, $rc, $mode);

    chmod 0664, $filename;
    print $fh "Hello\n";
    $fh->flush();

    $output = qx($system_config_test check_certfile $filename 2>&1);
    $rc = $? << 8;

    chomp($output);
    diag("Expect error: $output");

    $mode = (stat($filename))[2] & (S_IRWXU|S_IRWXG|S_IRWXO);

    unlink($filename);
    umask($umask);

    ok($rc != 0 && ($mode & (S_IRWXG|S_IRWXO)) != 0, "check_certfile_false");
}

sub get_cert_dir_env {
    my ($dirname) = File::Temp::tempdir(CLEANUP => 1);
    my ($output, $rc, $mode);
    my $old_dir = $ENV{X509_CERT_DIR};
    
    $ENV{X509_CERT_DIR} = $dirname;
    diag("X509_CERT_DIR=$dirname");

    $output = qx($system_config_test get_cert_dir);
    $rc = $? << 8;

    chomp($output);
    diag($output) if $output ne "";

    if ($old_dir)
    {
        $ENV{X509_CERT_DIR} = $old_dir;
    }
    else
    {
        delete $ENV{X509_CERT_DIR};
    }
    ok($dirname eq $output, "get_cert_dir_env");
}

sub get_cert_dir_env_format {
    my ($dirname) = File::Temp::tempdir(CLEANUP => 1);
    my $format_dirname = "$dirname/\%p";
    my ($output, $rc, $mode);
    my $old_dir = $ENV{X509_CERT_DIR};

    mkdir "$format_dirname";
    
    $ENV{X509_CERT_DIR} = $format_dirname;
    diag("X509_CERT_DIR=$format_dirname");

    $output = qx($system_config_test get_cert_dir);
    $rc = $? << 8;

    chomp($output);
    diag($output) if $output ne "";

    if ($old_dir)
    {
        $ENV{X509_CERT_DIR} = $old_dir;
    }
    else
    {
        delete $ENV{X509_CERT_DIR};
    }
    ok($rc == 0 && $format_dirname eq $output, "get_cert_dir_env_format");
}

sub get_cert_dir_env_bad {
    my ($dirname) = File::Temp::tempdir(CLEANUP => 0);
    my ($output, $rc, $mode);
    my $old_dir = $ENV{X509_CERT_DIR};

    rmdir $dirname;
    
    $ENV{X509_CERT_DIR} = $dirname;
    diag("X509_CERT_DIR=$dirname");

    $output = qx($system_config_test get_cert_dir 1>&2);
    $rc = $? << 8;

    chomp($output);
    diag($output) if $output ne "";

    if ($old_dir)
    {
        $ENV{X509_CERT_DIR} = $old_dir;
    }
    else
    {
        delete $ENV{X509_CERT_DIR};
    }
    ok($rc != 0, "get_cert_dir_env_bad");
}


sub get_cert_dir_home {
    my $homedir = (getpwuid($<))[7];
    my $certdir = "$homedir/.globus/certificates";
    my $old_dir = $ENV{X509_CERT_DIR};
    my ($rc, $output);
    
    SKIP: {
        skip 1, "User cert dir doesn't exist" unless -d $certdir;

        delete $ENV{X509_CERT_DIR} if $old_dir;

        $output = qx($system_config_test get_cert_dir);
        chomp($output);
        $rc = $? << 8;

        diag($output) if $output ne "";
        if ($old_dir) {
            $ENV{X509_CERT_DIR} = $old_dir;
        }

        ok($rc == 0 && $output eq $certdir, "get_cert_dir_home");
    }
}

sub get_user_cert_filename_pem {
    my $homedir = (getpwuid($<))[7];
    my $certfile = "$homedir/.globus/usercert.pem";
    my $keyfile = "$homedir/.globus/userkey.pem";
    my ($rc, $output);
    my ($testcertfile, $testkeyfile);
    my $old_cert_env = $ENV{X509_USER_CERT};
    my $old_key_env = $ENV{X509_USER_KEY};

    SKIP: {
        skip "Default PEM-formatted cert and key don't exist", 1
            unless -r $certfile && -r $keyfile;

        delete $ENV{X509_USER_CERT} if $old_cert_env;
        delete $ENV{X509_USER_KEY} if $old_key_env;
    
        $output = qx($system_config_test get_user_cert_filename);
        chomp($output);
        $rc = $? << 8;
        diag($output) if $output ne "";

        ($testcertfile, $testkeyfile) = split(/\n/, $output, 2);

        ok($rc == 0 && $testcertfile eq $certfile && $testkeyfile eq $keyfile,
           "get_user_cert_filename_pem");

        if ($old_cert_env) {
            $ENV{X509_USER_CERT} = $old_cert_env 
        }
        if ($old_key_env) {
            $ENV{X509_USER_KEY} = $old_key_env 
        }
    }
}

sub get_user_cert_filename_p12 {
    my $homedir = (getpwuid($<))[7];
    my $p12file = "$homedir/.globus/usercred.p12";
    my ($rc, $output);
    my ($testcertfile, $testkeyfile);
    my $old_cert_env = $ENV{X509_USER_CERT};
    my $old_key_env = $ENV{X509_USER_KEY};
    
    SKIP: {
        skip "Default PKCS12-formatted cert and key don't exist", 1
            unless -r $p12file;

        delete $ENV{X509_USER_CERT} if $old_cert_env;
        delete $ENV{X509_USER_KEY} if $old_key_env;
    
        $output = qx($system_config_test get_user_cert_filename);
        chomp($output);
        $rc = $? << 8;
        diag($output) if $output ne "";

        ($testcertfile, $testkeyfile) = split(/\n/, $output, 2);
        diag("testcertfile='$testcertfile'");
        diag("testkeyfile='$testkeyfile'");

        ok($rc == 0 && $testcertfile eq $p12file && $testkeyfile eq $p12file,
            "get_user_cert_filename_p12");
        if ($old_cert_env) {
            $ENV{X509_USER_CERT} = $old_cert_env 
        }
        if ($old_key_env) {
            $ENV{X509_USER_KEY} = $old_key_env 
        }
    }
}

sub get_user_cert_filename_env {
    my ($certfh, $certfile) = File::Temp::tempfile();
    my ($keyfh, $keyfile) = File::Temp::tempfile();
    my ($rc, $output);
    my ($testcertfile, $testkeyfile);
    my $old_cert_env = $ENV{X509_USER_CERT};
    my $old_key_env = $ENV{X509_USER_KEY};

    delete $ENV{X509_USER_CERT};
    delete $ENV{X509_USER_KEY};

    print $certfh "Hello\n";
    $certfh->flush();

    print $keyfh "Hello\n";
    $keyfh->flush();

    $ENV{X509_USER_CERT} = $certfile;
    $ENV{X509_USER_KEY} = $keyfile;


    $output = qx($system_config_test get_user_cert_filename);
    chomp($output);
    $rc = $? << 8;
    diag($output) if $output ne "";

    ($testcertfile, $testkeyfile) = split(/\n/, $output, 2);
    diag("testcertfile='$testcertfile'");
    diag("testkeyfile='$testkeyfile'");

    ok($rc == 0 && $testcertfile eq $certfile && $testkeyfile eq $keyfile,
            "get_user_cert_filename_env");

    if ($old_cert_env)
    {
        $ENV{X509_USER_CERT} = $old_cert_env;
    }
    else
    {
        delete $ENV{X509_USER_CERT};
    }
    if ($old_key_env)
    {
        $ENV{X509_USER_KEY} = $old_key_env;
    }
    else
    {
        delete $ENV{X509_USER_KEY};
    }
}

sub get_user_cert_filename_env_format {
    my $testdir = File::Temp::tempdir(CLEANUP => 1);
    my ($certfile) = "$testdir/cert%p";
    my ($keyfile) = "$testdir/key%p";
    my ($certfh, $keyfh);
    my ($rc, $output);
    my ($testcertfile, $testkeyfile);
    my $old_cert_env = $ENV{X509_USER_CERT};
    my $old_key_env = $ENV{X509_USER_KEY};

    delete $ENV{X509_USER_CERT};
    delete $ENV{X509_USER_KEY};

    $ENV{X509_USER_CERT} = $certfile;
    $ENV{X509_USER_KEY} = $keyfile;

    open $certfh, ">$certfile";
    print $certfh "Hello\n";
    close($certfh);
    chmod 0644, $certfile;

    open $keyfh, ">$keyfile";
    print $keyfh "Hello\n";
    close($keyfh);
    chmod 0600, $keyfile;

    $output = qx($system_config_test get_user_cert_filename);
    chomp($output);
    $rc = $? << 8;
    diag($output) if $output ne "";

    ($testcertfile, $testkeyfile) = split(/\n/, $output, 2);
    diag("testcertfile='$testcertfile'");
    diag("testkeyfile='$testkeyfile'");

    ok($rc == 0 && $testcertfile eq $certfile && $testkeyfile eq $keyfile,
            "get_user_cert_filename_env_format");

    if ($old_cert_env)
    {
        $ENV{X509_USER_CERT} = $old_cert_env;
    }
    else
    {
        delete $ENV{X509_USER_CERT};
    }
    if ($old_key_env)
    {
        $ENV{X509_USER_KEY} = $old_key_env;
    }
    else
    {
        delete $ENV{X509_USER_KEY};
    }
}

sub get_user_cert_filename_env_bad {
    my $testdir = File::Temp::tempdir(CLEANUP => 1);
    my ($certfile) = "$testdir/cert";
    my ($keyfile) = "$testdir/key";
    my ($rc, $output);
    my $old_cert_env = $ENV{X509_USER_CERT};
    my $old_key_env = $ENV{X509_USER_KEY};

    delete $ENV{X509_USER_CERT};
    delete $ENV{X509_USER_KEY};

    $ENV{X509_USER_CERT} = $certfile;
    $ENV{X509_USER_KEY} = $keyfile;

    $output = qx($system_config_test get_user_cert_filename 2>&1);
    chomp($output);
    $rc = $? << 8;
    diag($output) if $output ne "";

    ok($rc != 0, "get_user_cert_filename_env_bad");

    if ($old_cert_env)
    {
        $ENV{X509_USER_CERT} = $old_cert_env;
    }
    else
    {
        delete $ENV{X509_USER_CERT};
    }
    if ($old_key_env)
    {
        $ENV{X509_USER_KEY} = $old_key_env;
    }
    else
    {
        delete $ENV{X509_USER_KEY};
    }
}

sub get_vhost_cred_dir_globus_location
{
    my ($dirname) = File::Temp::tempdir(CLEANUP => 1);
    my ($output, $rc, $mode);
    my $old_dir = $ENV{X509_VHOST_CRED_DIR};
    my $old_globus_location = $ENV{GLOBUS_LOCATION};
    my $vhostdir = "$dirname/etc/vhosts";

    File::Path::make_path($vhostdir);

    delete $ENV{X509_VHOST_CRED_DIR};
    $ENV{GLOBUS_LOCATION} = $dirname;
    diag("GLOBUS_LOCATION=$dirname");

    $output = qx($system_config_test get_vhost_cred_dir);
    $rc = $? << 8;

    chomp($output);
    diag($output) if $output ne "";

    if ($old_dir)
    {
        $ENV{X509_VHOST_CRED_DIR} = $old_dir;
    }
    else
    {
        delete $ENV{X509_VHOST_CRED_DIR};
    }

    if ($old_dir) {
        $ENV{X509_VHOST_CRED_DIR} = $old_dir;
    }
    if ($old_globus_location) {
        $ENV{GLOBUS_LOCATION} = $old_globus_location;
    } else {
        delete $ENV{GLOBUS_LOCATION};
    }

    ok($rc == 0 && $output eq $vhostdir, "get_vhost_cred_dir_globus_location");
}

sub get_vhost_cred_dir_env
{
    my ($dirname) = File::Temp::tempdir(CLEANUP => 1);
    my ($output, $rc, $mode);
    my $old_dir = $ENV{X509_VHOST_CRED_DIR};
    
    $ENV{X509_VHOST_CRED_DIR} = $dirname;
    diag("X509_VHOST_CRED_DIR=$dirname");

    $output = qx($system_config_test get_vhost_cred_dir);
    $rc = $? << 8;

    chomp($output);
    diag($output) if $output ne "";

    if ($old_dir)
    {
        $ENV{X509_VHOST_CRED_DIR} = $old_dir;
    }
    else
    {
        delete $ENV{X509_VHOST_CRED_DIR};
    }
    ok($dirname eq $output, "get_vhost_cred_dir");
}

my @tests = qw (
    set_key_permissions
    get_home_dir
    file_exists_true
    file_exists_false
    dir_exists_true
    dir_exists_false
    check_keyfile_true
    check_keyfile_false
    check_certfile_true
    check_certfile_false
    get_cert_dir_env
    get_cert_dir_env_format
    get_cert_dir_env_bad
    get_cert_dir_home
    get_user_cert_filename_pem
    get_user_cert_filename_p12
    get_user_cert_filename_env
    get_user_cert_filename_env_format
    get_user_cert_filename_env_bad
    get_vhost_cred_dir_globus_location
    get_vhost_cred_dir_env
);

plan tests => scalar(@tests);

foreach (@tests) {
    eval $_;
    if ($@) {
        print STDERR $@;
    }
}
